src/Shell/Task/ModelTask.php

```diff
@@ -344,26 +351,69 @@ public function findHasMany($model, array $associations)
                 }
             }
 
-            foreach ($otherSchema->columns() as $fieldName) {
+            if ($tableName === $otherTableName && !in_array('parent_id', $otherSchema->columns())) {
+                continue;
+            }
+
+            $thisForeignKey = $primaryKey;
+            foreach ($thisForeignKey as $i => $fieldName) {
+                if ($fieldName === 'id') {
+                    if (in_array('parent_id', $schema->columns())
+                        && $tableName === $otherTableName)
+                    {
+                        $thisForeignKey[$i] = 'parent_id';
+                    } else {
+                        $thisForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
+                    }
+                }
+            }
+
+            $otherForeignKeys = $otherSchema->columns();
+            foreach ($otherForeignKeys as $i => $fieldName) {
+                if ($fieldName === 'parent_id' && $tableName !== $otherTableName) {
+                    unset($otherForeignKeys[$i]);
+                } elseif ($fieldName === 'id') {
+                    unset($otherForeignKeys[$i]);
+                } elseif (strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName)) {
+                    unset($otherForeignKeys[$i]);
+                } elseif (!in_array($fieldName, $thisForeignKey)) {
+                    unset($otherForeignKeys[$i]);
+                }
+            }
```

Comment:
this function is large enough to guarantee another Task class, can you move it there?